introduce protected_exprt

kroening opened the pull request:
This will serve as the interface that we aspire exprt to offer; in
particular, access to opX is protected.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: cd7d05d184b7669e6c8f7b0da18f0f8c0f3118fa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97934741
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

thk123 commented on src/pointer-analysis/value_set_fi.cpp:
```diff
@@ -870,8 +870,8 @@ void value_set_fit::get_reference_set_sharing_rec(
       {
-        index_exprt index_expr(
-          object, from_integer(0, index_type()), expr.type());
+        exprt index_expr =
```
:question: Why remove the type from lhs? Because we call `make_typecast` which makes this not an `index_exprt` :scream: -> :pick: don't change the type, instead create a typecast from the `index_exprt`

kroening replied:
#3856 will fix this!

thk123 commented on src/util/expr.h:
```diff
@@ -318,6 +318,40 @@ class exprt:public irept
+/// This API will eventually replace exprt.
+class protected_exprt : public exprt
+{
+protected:
```
:question: Would it make more sense to use protected inheritance? I assume the answer is no as there remains calls to other methods, that have not been removed, but perhaps given this direction of travel it would making more sense to `using` them into the public section? (I think you can do this, but I have not actually tried).

thk123 replied:
This would circumvent the need for this class entirely and just update the classes that inherit from `exprt` to protected extend it and in fact allow for explicitly identifying the classes that use the wrong interface.

kroening replied:
There are a lot of classes inheriting from `exprt`; furthermore, the interface of `exprt` should change as well, once we are there.

thk123 commented on src/util/expr.h:
```diff
@@ -318,6 +318,40 @@ class exprt:public irept
 
+/// Protected class for all expressions.
+/// This API will eventually replace exprt.
+class protected_exprt : public exprt
```
:pick: I slightly worry about the name, in general, `X_exprt` means an expression representing an AST node of type X. I guess `protected` is never a node of a C/C++/Java AST, but perhaps this should be called `expr_protected` or `expr_api_protected` to indicate that this is an internal fudge rather than an actual GOTO thing.

kroening replied:
ok

thk123 reviewed (commented):
This presumably requires a TG bump before merging - will create to see how much additional work is required.

thk123 commented:
Ah this doesn't yet compile on CBMC - will hold off on the bump front for now.

kroening committed:
avoid direct access to exprt::op0()

This will enable protecting exprt::opX()

kroening committed:
use conditional_cast instead of make_typecast

make_typecast isn't type safe; typecast_exprt::conditional_cast is.

kroening committed:
introduce expr_protectedt

This will serve as the interface that we aspire exprt to offer; in
particular, access to opX is protected.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5d451cc8212d424f82acfbc6a736b5d0dc432106).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98429525
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

thk123 commented:
@kroening This still isn't compiling on cbmc yet: 
```
value_set_fi.cpp:858:48: error: within this context
           index_expr.make_typecast(array.type());
```

So will hold off on the bump for now.

tautschnig commented on src/util/expr.h:
```diff
@@ -319,6 +319,40 @@ class exprt:public irept
   const_unique_depth_iteratort unique_depth_cend() const;
 };
 
+/// Protected class for all expressions.
```
What is a "protected" class? Who is being protected? From what risks are they being shielded? When should anyone use it? I think this comment needs to be reworded.

tautschnig replied:
I'm a bit confused, an intermittent version had more elaborate documentation. Do a `git reflog` to look at those?

tautschnig commented on src/util/expr.h:
```diff
@@ -319,6 +319,40 @@ class exprt:public irept
+  DEPRECATED("use protected_exprt(id, type) instead")
+  explicit protected_exprt(const irep_idt &_id) : exprt(_id)
+  {
+  }
```
It seems rather weird that newly introduced code would be deprecated? Can we just not have these?

kroening replied:
sure, removed

tautschnig replied:
Back again?

tautschnig commented on src/util/expr.h:
```diff
@@ -319,6 +319,40 @@ class exprt:public irept
+  using exprt::op3;
+  // using exprt::find;
+  // using exprt::get;
+  // using exprt::set;
```
Adding commented-out code without comments is just raising questions, but not providing answers.

kroening replied:
sure, removed

tautschnig commented on src/util/pointer_offset_size.cpp:
```diff
@@ -593,8 +593,11 @@ exprt build_sizeof_expr(
-  const typet &type=
-    static_cast<const typet &>(expr.find(ID_C_c_sizeof_type));
+  // need to cast down to access 'find'
+  const auto &expr_as_irep = static_cast<const irept &>(expr);
```
I don't think that's an improvement?

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 7b6444ad7f1c7a74b2413cd8fbf39251ab54e8a6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98484788
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: d9ae0f752817912ddfc5b7f72a8208400306e31c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98523776
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

tautschnig commented on src/util/pointer_offset_size.cpp:
```diff
@@ -593,8 +593,11 @@ exprt build_sizeof_expr(
+  const auto &expr_as_irep = static_cast<const irept &>(expr);
+
+  const typet &type =
+    static_cast<const typet &>(expr_as_irep.find(ID_C_c_sizeof_type));
```
Is this still necessary, given `.find` is no longer protected?

kroening replied:
removed

tautschnig commented on src/util/std_expr.h:
```diff
@@ -969,11 +962,17 @@ class multi_ary_exprt:public exprt
+  using exprt::op0;
+  using exprt::op1;
+  using exprt::op2;
+  using exprt::op3;
```
Should this really be done? A `multi_ary_exprt` can, but need not have this many operands? For an unidentified `multi_ary_exprt` the caller should always use `.operands()` and just operate on that data structure.

kroening replied:
replaced by ranged-checked opX().

hannes-steffenhagen-diffblue commented on src/util/expr.h:
```diff
@@ -319,6 +319,29 @@ class exprt:public irept
 
+/// Class for expressions that offers a restriction of the API
+/// offered by exprt. Notably, access to non-type-safe methods such
+/// as opX() is protected.
```
Does that mean that `exprt` subclasses that so far _only_ have these as their API will get updated (e.g. classes like `typecast_exprt` that only have `op0` rather than a named accessor?).

Also, if we're already going to do an API do-over, why make the "new" expr inherit from `irept` at all, rather than have an irep as a protected field or something?

hannes-steffenhagen-diffblue reviewed (commented):
:thinking:

kroening committed:
multi_ary_exprt::opX with precondtion

This prevents out-of-bound accesses to operands().

tautschnig reviewed (changes requested):
It looks like some earlier version has been restored? Might be worth looking at the commits the GitHub lists or going for `git reflog`.

kroening committed:
replace exprt::make_typecast() by typecast_exprt()

exprt::make_typecast() isn't type safe

kroening commented:
Back again!

tautschnig commented on src/util/std_expr.h:
```diff
@@ -753,6 +746,10 @@ class binary_exprt:public exprt
+
   const exprt &op2() const = delete;
   exprt &op2() = delete;
   const exprt &op3() const = delete;
```
Is this still needed?

kroening replied:
Yes, there are still too many users -- but I'd say that it makes sense to work towards removing these two lines.

kroening replied:
(But at least it's not dangerous to use them).

tautschnig replied:
I don't understand - these are being deleted, even though `protected_exprt` makes them unavailable anyway.

kroening replied:
Ah, ok, the "this" was unclear.

The deletion of op2() and op3() is indeed no longer needed.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: a48f855762bb37d2a510ceddb1ee6d5939b45e2e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98702385
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

martin-cs reviewed (approved):
The changes to bits of code I am responsible for are fine.  I will leave it to others to have opinions on the exact API.  Being cleaner about the use of exprts / irepts is a good thing in general.

peterschrammel commented on src/util/expr.h:
```diff
@@ -319,6 +319,28 @@ class exprt:public irept
 };
 
+/// Base class for all expressions. This protects low-level methods in
+/// exprt that are not type safe. Depcrecated constructors are removed.
```
Typo in `deprecated`
